File: OneCodeTeam/How to use SqlDependency to get the notification in Entity Framework/[C#]-How to use SqlDependency to get the notification in Entity Framework/C#/CSEFAutoUpdate/AutoUpdate.cs
/****************************** Module Header ******************************\
Module Name:  AutoUpdate.cs
Project:      CSEFAutoUpdate
Copyright (c) Microsoft Corporation.

We can use the Sqldependency to get the notification when the data is changed 
in database, but EF doesn’t have the same feature. In this sample, we will 
demonstrate how to automatically update by Sqldependency in Entity Framework.
In this sample, we will demonstrate two ways that use SqlDependency to get the 
change notification to auto update data:
1. Get the notification of changes immediately.
2. Get the notification of changes regularly.

This source is subject to the Microsoft Public License.
See http://www.microsoft.com/en-us/openness/licenses.aspx#MPL.
All other rights reserved.

THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED 
WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
\***************************************************************************/

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data.Entity.Infrastructure;
using System.Data.Objects;
using System.Data.SqlClient;
using System.Linq;
using System.Windows.Forms;
using System.Security.Permissions;
using System.Security;

namespace CSEFAutoUpdate
{
    public partial class AutoUpdate : Form
    {
        private WarehouseContext warehouse = new WarehouseContext();
        private IQueryable iquery = null;
        private ImmediateNotificationRegister<Product> notification = null;
        private RegularlyNotificationRegister<Product> regularNotificaton = null;
        private Timer formTimer = null;
        private Int32 interval;
        private Int32 count;

        private bool CanRequestNotifications()
        {
            // In order to use the callback feature of the
            // SqlDependency, the application must have
            // the SqlClientPermission permission.
            try
            {
                SqlClientPermission perm =
                    new SqlClientPermission(
                    PermissionState.Unrestricted);

                perm.Demand();

                return true;
            }
            catch (SecurityException se)
            {
                MessageBox.Show(se.Message, "Permission Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
                return false;
            }
            catch (Exception e)
            {
                MessageBox.Show(e.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
                return false;
        }
        }

        public AutoUpdate()
        {
            InitializeComponent();
        }

        private void AutoUpdate_Load(object sender, EventArgs e)
        {
            btnGetData.Enabled = CanRequestNotifications();

            CreateDatabaseIfNotExist();

            // start immediate notification register
            ImmediateNotificationRegister<Product>.StartMonitor(warehouse);
            RegularlyNotificationRegister<Product>.StartMonitor(warehouse);
        }


        /// <summary>
        /// Create the database and insert the data if there's no database
        /// </summary>
        private void CreateDatabaseIfNotExist()
        {
            if (warehouse != null && !warehouse.Database.Exists())
            {
                Product[] products = 
                { 
                  new Product { Name = "Red Bicycle",Price=(Decimal)805.5,Amount=1050 },
                  new Product { Name = "White Bicycle",Price=(Decimal)1049.9,Amount=312 },
                  new Product { Name = "Black Bicycle",Price=(Decimal)888.8,Amount=965 }
                };

                foreach (Product product in products)
                {
                    warehouse.Products.Add(product);
                }

                warehouse.SaveChanges();
            }
        }

        private void GetData_Click(object sender, EventArgs e)
        {
            Decimal lowPrice, highPrice;

            if (!Decimal.TryParse(txtLowPrice.Text, out lowPrice))
            {
                lowPrice = 0;
            }

            if (!Decimal.TryParse(txtHighPrice.Text, out highPrice))
            {
                highPrice = Decimal.MaxValue;
            }


            // Create the query.
            iquery = from p in warehouse.Products
                     where p.Price >= lowPrice && p.Price <= highPrice
                     select p;

            if (this.rabtnImUpdate.Checked)
            {
                // If need to update immediately, use ImmediateNotificationRegister to register 
                // SqlDependency.
                notification = new ImmediateNotificationRegister<Product>(warehouse, iquery);
                notification.OnChanged += NotificationOnChanged;
            }
            else
            {
                // We can use RegularlyNotificationRegister to implement update regularly.
                if (Int32.TryParse(this.txtInterval.Text, out interval))
                {
                    regularNotificaton = new RegularlyNotificationRegister<Product>(warehouse, iquery, interval * 1000);
                    regularNotificaton.OnChanged += NotificationOnChanged;

                    // Only for displaying the progress
                    this.proBar.Value = 100 / interval;
                    count = 1;
                    formTimer = new Timer();
                    formTimer.Interval = 1000;
                    formTimer.Tick += formTimer_Tick;
                    formTimer.Start();
                }
                else
                {
                    return;
                }
            }

            GetData();

            ChangeButtonState();
        }

        /// <summary>
        /// Display progress
        /// </summary>
        void formTimer_Tick(object sender, EventArgs e)
        {
            count = (count) % interval + 1;
            this.proBar.Value = (100 * count) / interval;
        }

        /// <summary>
        /// When changed the data, the method will be invoked.
        /// </summary>
        void NotificationOnChanged(object sender, EventArgs e)
        {
            // If InvokeRequired returns True, the code
            // is executing on a worker thread.
            if (this.InvokeRequired)
            {
                // Create a delegate to perform the thread switch.
                this.BeginInvoke(new Action(GetData), null);

                return;
            }

            GetData();
        }

        /// <summary>
        /// Display the data in the DataGridView
        /// </summary>
        void GetData()
        {
            dgvWatch.DataSource = (iquery as DbQuery<Product>).ToList();
        }

        /// <summary>
        /// Update the new price.
        /// </summary>
        private void Update_Click(object sender, EventArgs e)
        {
            Int32 id;
            Decimal newPrice;

            if (Int32.TryParse(txtId.Text, out id) && Decimal.TryParse(txtNewPrice.Text, out newPrice))
            {
                Product product = (from p in warehouse.Products
                                   where p.Id == id
                                   select p).FirstOrDefault();

                if (product != null)
                {
                    product.Price = newPrice;
                    warehouse.SaveChanges();
                }
            }
            else
            {
                MessageBox.Show("Please input the valid values.");
            }
        }


        private void dgvWatch_CellMouseClick(object sender, DataGridViewCellMouseEventArgs e)
        {
            DataGridView datagrid = (DataGridView)sender;
            List<Product> products = (List<Product>)datagrid.DataSource;

            if (e.RowIndex >= 0 && e.RowIndex < products.Count)
            {
                Product product = products[e.RowIndex];

                txtId.Text = product.Id.ToString();
                txtNewPrice.Text = product.Price.ToString();
            }
        }

        /// <summary>
        /// Stop SqlDependency.
        /// </summary>
        private void StopSqlDependency(object sender, EventArgs e)
        {
            try
            {
                if (notification != null)
                {
                    notification.Dispose();
                    notification = null;
                }

                if (regularNotificaton != null)
                {
                    regularNotificaton.Dispose();
                    regularNotificaton = null;
                }

                if (formTimer != null)
                {
                    formTimer.Stop();
                    formTimer.Dispose();
                    formTimer = null;
                    this.proBar.Value = 0;
                }

                ChangeButtonState();
            }
            catch (ArgumentException ex)
            {
                MessageBox.Show(ex.Message, "Paramter Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
            }
            catch (Exception ex)
            {
                if (ex.InnerException != null)
                {
                    MessageBox.Show(ex.Message + "(" + ex.InnerException.Message + ")", "Failed to Stop SqlDependency", MessageBoxButtons.OK, MessageBoxIcon.Error);
                }
                else
                {
                    MessageBox.Show(ex.Message, "Failed to Stop SqlDependency", MessageBoxButtons.OK, MessageBoxIcon.Error);
                }
            }
        }

        private void ChangeButtonState()
        {
            Boolean state = !this.btnGetData.Enabled;

            this.btnGetData.Enabled = state;
            this.txtLowPrice.Enabled = state;
            this.txtHighPrice.Enabled = state;
            this.rabtnImUpdate.Enabled = state;
            this.rabtnRegUpdate.Enabled = state;
            ChangeStateByRadioButton();

            this.txtId.Enabled = !state;
            this.txtNewPrice.Enabled = !state;
            this.btnUpdate.Enabled = !state;

            this.btnStop.Enabled = !state;
        }

        private void ChangeStateByRadioButton()
        {
            Boolean state = this.rabtnRegUpdate.Enabled && this.rabtnRegUpdate.Checked;

            this.txtInterval.Enabled = state;
            this.proBar.Enabled = state;
        }

        private void rabtnRegUpdate_CheckedChanged(object sender, EventArgs e)
        {
            ChangeStateByRadioButton();
        }

        private void AutoUpdate_FormClosed(object sender, FormClosedEventArgs e)
        {
            // stop notification monitor
            ImmediateNotificationRegister<Product>.StopMonitor(warehouse);
            RegularlyNotificationRegister<Product>.StopMonitor(warehouse);
        }
    }
}

File: OneCodeTeam/How to use SqlDependency to get the notification in Entity Framework 2/[VB]-How to use SqlDependency to get the notification in Entity Framework/VB/VBEFAutoUpdate/AutoUpdate.vb
'**************************** Module Header ******************************\
' Module Name:  AutoUpdate.vb
' Project:      VBEFAutoUpdate
' Copyright (c) Microsoft Corporation.
' 
' We can use the Sqldependency to get the notification when the data is changed 
' in database, but EF doesn’t have the same feature. In this sample, we will 
' demonstrate how to automatically update by Sqldependency in Entity Framework.
' In this sample, we will demonstrate two ways that use SqlDependency to get the 
' change notification to auto update data:
' 1. Get the notification of changes immediately.
' 2. Get the notification of changes regularly.
' 
' This source is subject to the Microsoft Public License.
' See http://www.microsoft.com/en-us/openness/licenses.aspx#MPL.
' All other rights reserved.
' 
' THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
' EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED 
' WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
'**************************************************************************/

Imports System.ComponentModel
Imports System.Data.Entity.Infrastructure
Imports System.Data.Objects
Imports System.Data.SqlClient
Imports System.Linq
Imports System.Security.Permissions
Imports System.Security

Partial Public Class AutoUpdate
    Inherits Form
    Private warehouse As New WarehouseContext()
    Private iquery As IQueryable = Nothing
    Private notification As ImmediateNotificationRegister(Of Product) = Nothing
    Private regularNotificaton As RegularlyNotificationRegister(Of Product) = Nothing
    Private formTimer As Timer = Nothing
    Private interval As Int32
    Private count As Int32

    Private Function CanRequestNotifications() As Boolean
        ' In order to use the callback feature of the
        ' SqlDependency, the application must have
        ' the SqlClientPermission permission.
        Try
            Dim perm As New SqlClientPermission(PermissionState.Unrestricted)

            perm.Demand()

            Return True
        Catch se As SecurityException
            MessageBox.Show(se.Message, "Permission Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return False
        Catch e As Exception
            MessageBox.Show(e.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return False
        End Try
    End Function

    Public Sub New()
        InitializeComponent()
    End Sub

    Private Sub AutoUpdate_Load(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
        btnGetData.Enabled = CanRequestNotifications()

        CreateDatabaseIfNotExist()

        ' start notification monitor
        ImmediateNotificationRegister(Of Product).StartMonitor(warehouse)
        RegularlyNotificationRegister(Of Product).StartMonitor(warehouse)
    End Sub


    ''' <summary>
    ''' Create the database and insert the data if there's no database
    ''' </summary>
    Private Sub CreateDatabaseIfNotExist()
        If warehouse IsNot Nothing AndAlso (Not warehouse.Database.Exists()) Then
            Dim products() As Product = {
                New Product With {.Name = "Red Bicycle", .Price = CType(805.5, Decimal), .Amount = 1050},
                New Product With {.Name = "White Bicycle", .Price = CType(1049.9, Decimal), .Amount = 312},
                New Product With {.Name = "Black Bicycle", .Price = CType(888.8, Decimal), .Amount = 965}}

            For Each product As Product In products
                warehouse.Products.Add(product)
            Next product

            warehouse.SaveChanges()
        End If
    End Sub

    Private Sub GetData_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnGetData.Click
        Dim lowPrice, highPrice As Decimal

        If Not Decimal.TryParse(txtLowPrice.Text, lowPrice) Then
            lowPrice = 0
        End If

        If Not Decimal.TryParse(txtHighPrice.Text, highPrice) Then
            highPrice = Decimal.MaxValue
        End If


        ' Create the query.
        iquery = From p In warehouse.Products
                 Where p.Price >= lowPrice AndAlso p.Price <= highPrice
                 Select p

        If Me.rabtnImUpdate.Checked Then
            ' If need to update immediately, use ImmediateNotificationRegister to register 
            ' SqlDependency.
            notification = New ImmediateNotificationRegister(Of Product)(warehouse, iquery)
            AddHandler notification.OnChanged, AddressOf NotificationOnChanged
        Else
            ' We can use RegularlyNotificationRegister to implement update regularly.
            If Int32.TryParse(Me.txtInterval.Text, interval) Then
                regularNotificaton = New RegularlyNotificationRegister(Of Product)(warehouse, iquery, interval * 1000)
                AddHandler regularNotificaton.OnChanged, AddressOf NotificationOnChanged

                ' Only for displaying the progress
                Me.proBar.Value = 100 \ interval
                count = 1
                formTimer = New Timer()
                formTimer.Interval = 1000
                AddHandler formTimer.Tick, AddressOf formTimer_Tick
                formTimer.Start()
            Else
                Return
            End If
        End If

        GetData()

        ChangeButtonState()
    End Sub

    ''' <summary>
    ''' Display progress
    ''' </summary>
    Private Sub formTimer_Tick(ByVal sender As Object, ByVal e As EventArgs)
        count = count Mod interval + 1
        Me.proBar.Value = (100 * count) / interval
    End Sub

    ''' <summary>
    ''' When changed the data, the method weill be invokced.
    ''' </summary>
		Private Sub NotificationOnChanged(ByVal sender As Object, ByVal e As EventArgs)
        ' If InvokeRequired returns True, the code
        ' is executing on a worker thread.
        If Me.InvokeRequired Then
            ' Create a delegate to perform the thread switch.
            Me.BeginInvoke(New Action(AddressOf GetData), Nothing)

            Return
        End If

        GetData()
    End Sub

    ''' <summary>
    ''' Display the data in the DataGridView
    ''' </summary>
    Private Sub GetData()
        dgvWatch.DataSource = (TryCast(iquery, DbQuery(Of Product))).ToList()
    End Sub

    ''' <summary>
    ''' Update the new price.
    ''' </summary>
    Private Sub Update_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnUpdate.Click
        Dim id As Int32
        Dim newPrice As Decimal

        If Int32.TryParse(txtId.Text, id) AndAlso Decimal.TryParse(txtNewPrice.Text, newPrice) Then
            Dim product As Product = (
                From p In warehouse.Products
                Where p.Id = id
                Select p).FirstOrDefault()

            If product IsNot Nothing Then
                product.Price = newPrice
                warehouse.SaveChanges()
            End If
        Else
            MessageBox.Show("Please input the valid values.")
        End If
    End Sub


    Private Sub dgvWatch_CellMouseClick(ByVal sender As Object,
                                        ByVal e As DataGridViewCellMouseEventArgs) Handles dgvWatch.CellMouseClick
        Dim datagrid As DataGridView = CType(sender, DataGridView)
        Dim products As List(Of Product) = CType(datagrid.DataSource, List(Of Product))

        If e.RowIndex >= 0 AndAlso e.RowIndex < products.Count Then
            Dim product As Product = products(e.RowIndex)

            txtId.Text = product.Id.ToString()
            txtNewPrice.Text = product.Price.ToString()
        End If
    End Sub

    ''' <summary>
    ''' Stop SqlDependency.
    ''' </summary>
    Private Sub StopSqlDependency(ByVal sender As Object, ByVal e As EventArgs) Handles btnStop.Click
        Try
            If notification IsNot Nothing Then
                notification.Dispose()
                notification = Nothing
            End If

            If regularNotificaton IsNot Nothing Then
                regularNotificaton.Dispose()
                regularNotificaton = Nothing
            End If

            If formTimer IsNot Nothing Then
                formTimer.Stop()
                formTimer.Dispose()
                formTimer = Nothing
                Me.proBar.Value = 0
            End If

            ChangeButtonState()
        Catch ex As ArgumentException
            MessageBox.Show(ex.Message, "Paramter Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
        Catch ex As Exception
            If ex.InnerException IsNot Nothing Then
                MessageBox.Show(ex.Message & "(" & ex.InnerException.Message & ")", "Failed to Stop SqlDependency", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Else
                MessageBox.Show(ex.Message, "Failed to Stop SqlDependency", MessageBoxButtons.OK, MessageBoxIcon.Error)
            End If
        End Try
    End Sub

    Private Sub ChangeButtonState()
        Dim state As Boolean = Not Me.btnGetData.Enabled

        Me.btnGetData.Enabled = state
        Me.txtLowPrice.Enabled = state
        Me.txtHighPrice.Enabled = state
        Me.rabtnImUpdate.Enabled = state
        Me.rabtnRegUpdate.Enabled = state
        ChangeStateByRadioButton()

        Me.txtId.Enabled = Not state
        Me.txtNewPrice.Enabled = Not state
        Me.btnUpdate.Enabled = Not state

        Me.btnStop.Enabled = Not state
    End Sub

    Private Sub ChangeStateByRadioButton()
        Dim state As Boolean = Me.rabtnRegUpdate.Enabled AndAlso Me.rabtnRegUpdate.Checked

        Me.txtInterval.Enabled = state
        Me.proBar.Enabled = state
    End Sub

    Private Sub rabtnRegUpdate_CheckedChanged(ByVal sender As Object, ByVal e As EventArgs) Handles rabtnRegUpdate.CheckedChanged
        ChangeStateByRadioButton()
    End Sub

    Private Sub AutoUpdate_FormClosed(sender As Object, e As FormClosedEventArgs) Handles MyBase.FormClosed
        ' stop notification monitor
        ImmediateNotificationRegister(Of Product).StopMonitor(warehouse)
        RegularlyNotificationRegister(Of Product).StopMonitor(warehouse)
    End Sub
End Class

